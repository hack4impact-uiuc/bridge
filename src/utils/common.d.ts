declare module "@hack4impact-uiuc/bridge/common" {
  import { Props, ReactType } from "react";
  import {
    ColorProps,
    SpaceProps,
    LayoutProps,
    TypographyProps as StyledTypographyProps,
    BordersProps as StyledBordersProps,
    BoxShadowProps,
  } from "styled-system";

  export interface BaseProps extends Props<any> {
    as?: ReactType;
    className?: string;
  }
  export interface CommonProps
    extends BaseProps,
      ColorProps,
      SpaceProps,
      LayoutProps {}

  export interface TypographyProps extends BaseProps, StyledTypographyProps {}

  export interface BorderProps extends StyledBordersProps, BoxShadowProps {}
}
