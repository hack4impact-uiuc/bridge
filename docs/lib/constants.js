export const COMPONENTS_URL = '/components';
export const DESIGN_URL = '/design';
