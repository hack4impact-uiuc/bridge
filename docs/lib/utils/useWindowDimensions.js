import { useState, useEffect } from 'react';

function getWindowDimensions() {
  const { innerWidth: width, innerHeight: height } = window;
  return {
    width,
    height,
  };
}

export default function useWindowDimensions() {
  const [windowDimensions, setWindowDimensions] = useState(getWindowDimensions());

  useEffect(() => {
    function handleResize() {
      const { innerWidth: width, innerHeight: height } = window;
      setWindowDimensions({ width, height });
    }

    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return windowDimensions;
}
