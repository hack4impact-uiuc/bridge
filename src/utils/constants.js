import * as styledSystem from 'styled-system';
import systemPropTypes from '@styled-system/prop-types';
import PropTypes from 'prop-types';


const { compose, system } = styledSystem;

// Text Decoration Prop
export const textDecoration = system({
  prop: 'textDecoration',
  cssProperty: 'text-decoration',
});
textDecoration.propTypes = PropTypes.string;


// common style props to be passed into components
export const COMMON = compose(styledSystem.space, styledSystem.layout);
COMMON.propTypes = {
  ...systemPropTypes.space,
  ...systemPropTypes.layout,
};

// for easier access
export const TYPOGRAPHY = styledSystem.typography;
export const LAYOUT = styledSystem.layout;
export const POSITION = styledSystem.position;
export const FLEX = styledSystem.flexbox;
export const GRID = styledSystem.grid;
export const COLOR = styledSystem.color;
export const BORDER = styledSystem.border;

TYPOGRAPHY.propTypes = systemPropTypes.typography;
LAYOUT.propTypes = systemPropTypes.layout;
POSITION.propTypes = systemPropTypes.position;
FLEX.propTypes = systemPropTypes.flexbox;
GRID.propTypes = systemPropTypes.grid;
COLOR.propTypes = systemPropTypes.color;
BORDER.propTypes = systemPropTypes.border;
