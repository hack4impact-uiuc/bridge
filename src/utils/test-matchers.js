/* eslint-disable react/forbid-foreign-prop-types */
// taken from https://github.com/primer/components/blob/master/src/utils/test-matchers.js
import 'jest-styled-components';

const stringify = (d) => JSON.stringify(d, null, '  ');
/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap'];

// eslint-disable-next-line no-undef
expect.extend({
  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes));
    const expectedPropKeys = Object.keys(propNames.propTypes);
    const missing = expectedPropKeys.filter((key) => !propKeys.has(key))
      .filter((key) => !ALIAS_PROP_TYPES.includes(key));
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`,
    };
  },
});
