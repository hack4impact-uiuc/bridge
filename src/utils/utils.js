import themeGet from '@styled-system/theme-get';
import * as styledSystem from 'styled-system';
import theme from '../theme';

const { get: getKey } = styledSystem;

export const get = (key) => themeGet(key, getKey(theme, key));

export const getOrDefault = (val, def) => (val !== undefined && val !== null ? val : def);

export const lodashGet = (object, path, defaultVal) => {
  const PATH = Array.isArray(path)
    ? path
    : path.split('.').filter((i) => i.length);
  if (!PATH.length) {
    return object === undefined ? defaultVal : object;
  }
  if (object === null || object === undefined || typeof (object[PATH[0]]) === 'undefined') {
    return defaultVal;
  }
  return lodashGet(object[PATH.shift()], PATH, defaultVal);
};
