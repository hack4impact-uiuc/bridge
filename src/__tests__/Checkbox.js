import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Checkbox } from '..';
import { COMMON, FLEX, BORDER } from '../utils/constants';
import { renderJSON } from '../utils/testing';
import 'babel-polyfill'; // axe violations required babel-polyfill

expect.extend(toHaveNoViolations);

describe('Checkbox', () => {
  it('renders a <label>', () => {
    expect(renderJSON(<Checkbox />).type).toEqual('label');
  });

  it('should have no axe violations', async () => {
    const { container } = render(
      <Checkbox id="option" value="option">
        option
      </Checkbox>,
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    cleanup();
  });

  it('implements system props', () => {
    expect(Checkbox).toImplementSystemProps(COMMON);
  });

  it('renders children', () => {
    const CheckboxText = 'option';
    render(<Checkbox>{CheckboxText}</Checkbox>);
    expect(screen.getByText(CheckboxText)).toBeInTheDocument();
  });

  it('respects "width" props', () => {
    const { container } = render(<Checkbox width={200} />);
    expect(container.firstChild).toHaveStyle('width: 200px;');
  });

  it('respects the "disabled" prop', () => {
    const item = renderJSON(<Checkbox disabled>Checkbox</Checkbox>);
    expect(item.props.disabled).toEqual(true);
    expect(item).toMatchSnapshot();
  });

  it('respects "error" prop', () => {
    const item = renderJSON(<Checkbox error="error">Checkbox</Checkbox>);
    expect(item).toMatchSnapshot();
  });
});

describe('Checkbox Group', () => {
  it('implements system props - same as Box', () => {
    // same as box
    expect(Checkbox.Group).toImplementSystemProps(COMMON);
    expect(Checkbox.Group).toImplementSystemProps(FLEX);
    expect(Checkbox.Group).toImplementSystemProps(BORDER);
  });

  it('renders correctly with 28px left margin per child besides first child', () => {
    expect(
      renderJSON(
        <Checkbox.Group>
          <Checkbox id="1">option 1</Checkbox>
          <Checkbox id="2">option 2</Checkbox>
          <Checkbox id="3">option 3</Checkbox>
        </Checkbox.Group>,
      ),
    ).toMatchSnapshot();
  });

  it('passes down "disabled" prop to children', () => {
    const group = renderJSON(
      <Checkbox.Group disabled>
        <Checkbox id="1">option 1</Checkbox>
        <Checkbox id="2">option 2</Checkbox>
        <Checkbox id="3">option 3</Checkbox>
      </Checkbox.Group>,
    );
    group.children.map((child) => expect(child.props.disabled).toBeTruthy());
  });
});
