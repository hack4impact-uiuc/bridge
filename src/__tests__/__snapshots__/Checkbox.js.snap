// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Group renders correctly with 28px left margin per child besides first child 1`] = `
.c0 {
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 18px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 20px;
  color: #272A2D;
  position: relative;
  padding-left: 40px;
  cursor: pointer;
  margin-left: 0px;
}

.c0 input:checked ~ span {
  background: #155DA1;
  border: 1px solid #155DA1;
  border-radius: 3px;
  display: block;
}

.c0 input:checked ~ span svg {
  display: inline-block;
}

.c3 {
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 18px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 20px;
  color: #272A2D;
  position: relative;
  padding-left: 40px;
  cursor: pointer;
  margin-left: 28px;
}

.c3 input:checked ~ span {
  background: #155DA1;
  border: 1px solid #155DA1;
  border-radius: 3px;
  display: block;
}

.c3 input:checked ~ span svg {
  display: inline-block;
}

.c1 {
  position: absolute;
  opacity: 0;
  height: 0;
  width: 0;
}

.c2 {
  position: absolute;
  top: 0;
  left: 0;
  height: 18px;
  width: 18px;
  background: #FFFFFF;
  border: 1px solid #8B9199;
  border-radius: 3px;
}

.c2::after {
  content: "";
  position: absolute;
  display: none;
}

.c2 svg {
  position: absolute;
  top: 4px;
  left: 2.2px;
  height: 10px;
  display: none;
}

.c2[disabled] {
  background: #EBEEF2;
  border: 1px solid #666B72;
}

<div
  className=""
>
  <label
    className="c0"
    htmlFor="1"
  >
    option 1
    <input
      className="c1"
      defaultChecked={true}
      id="1"
      type="checkbox"
    />
    <span
      className="c2"
    >
      <svg
        fill="none"
        height="10"
        viewBox="0 0 14 10"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13 1L4.75 9L1 5.36364"
          stroke="white"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        />
      </svg>
    </span>
  </label>
  <label
    className="c3"
    htmlFor="2"
  >
    option 2
    <input
      className="c1"
      defaultChecked={true}
      id="2"
      type="checkbox"
    />
    <span
      className="c2"
    >
      <svg
        fill="none"
        height="10"
        viewBox="0 0 14 10"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13 1L4.75 9L1 5.36364"
          stroke="white"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        />
      </svg>
    </span>
  </label>
  <label
    className="c3"
    htmlFor="3"
  >
    option 3
    <input
      className="c1"
      defaultChecked={true}
      id="3"
      type="checkbox"
    />
    <span
      className="c2"
    >
      <svg
        fill="none"
        height="10"
        viewBox="0 0 14 10"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13 1L4.75 9L1 5.36364"
          stroke="white"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
        />
      </svg>
    </span>
  </label>
</div>
`;

exports[`Checkbox respects "error" prop 1`] = `
.c0 {
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 18px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 20px;
  color: #272A2D;
  position: relative;
  padding-left: 40px;
  cursor: pointer;
}

.c0 input:checked ~ span {
  background: #A62B31;
  border: 1px solid #A62B31;
  border-radius: 3px;
  display: block;
}

.c0 input:checked ~ span svg {
  display: inline-block;
}

.c1 {
  position: absolute;
  opacity: 0;
  height: 0;
  width: 0;
}

.c2 {
  position: absolute;
  top: 0;
  left: 0;
  height: 18px;
  width: 18px;
  background: #FFFFFF;
  border: 1px solid #8B9199;
  border-radius: 3px;
}

.c2::after {
  content: "";
  position: absolute;
  display: none;
}

.c2 svg {
  position: absolute;
  top: 4px;
  left: 2.2px;
  height: 10px;
  display: none;
}

.c2[disabled] {
  background: #EBEEF2;
  border: 1px solid #666B72;
}

<label
  className="c0"
>
  Checkbox
  <input
    className="c1"
    type="checkbox"
  />
  <span
    className="c2"
  >
    <svg
      fill="none"
      height="10"
      viewBox="0 0 14 10"
      width="14"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M13 1L4.75 9L1 5.36364"
        stroke="white"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
      />
    </svg>
  </span>
</label>
`;

exports[`Checkbox respects the "disabled" prop 1`] = `
.c0 {
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 18px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 20px;
  color: #272A2D;
  position: relative;
  padding-left: 40px;
  cursor: pointer;
}

.c0 input:checked ~ span {
  background: #155DA1;
  border: 1px solid #155DA1;
  border-radius: 3px;
  display: block;
}

.c0 input:checked ~ span svg {
  display: inline-block;
}

.c1 {
  position: absolute;
  opacity: 0;
  height: 0;
  width: 0;
}

.c2 {
  position: absolute;
  top: 0;
  left: 0;
  height: 18px;
  width: 18px;
  background: #FFFFFF;
  border: 1px solid #8B9199;
  border-radius: 3px;
}

.c2::after {
  content: "";
  position: absolute;
  display: none;
}

.c2 svg {
  position: absolute;
  top: 4px;
  left: 2.2px;
  height: 10px;
  display: none;
}

.c2[disabled] {
  background: #EBEEF2;
  border: 1px solid #666B72;
}

<label
  className="c0"
  disabled={true}
>
  Checkbox
  <input
    className="c1"
    disabled={true}
    type="checkbox"
  />
  <span
    className="c2"
    disabled={true}
  >
    <svg
      fill="none"
      height="10"
      viewBox="0 0 14 10"
      width="14"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M13 1L4.75 9L1 5.36364"
        stroke="white"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
      />
    </svg>
  </span>
</label>
`;
